src/optimagic/optimizers/bayesian_optimizer.py

```diff
@@ -0,0 +1,347 @@
+
+            bounds_transformer = SequentialDomainReductionTransformer(
+                gamma_osc=self.sdr_gamma_osc,
+                gamma_pan=self.sdr_gamma_pan,
+                eta=self.sdr_eta,
+                minimum_window=self.sdr_minimum_window,
+            )
+
+        optimizer = BayesianOptimization(
+            f=objective,
+            pbounds=pbounds,
+            acquisition_function=acq,
+            constraint=constraint,
+            random_state=self.random_state,
+            verbose=self.verbose,
+            bounds_transformer=bounds_transformer,
+            allow_duplicate_points=self.allow_duplicate_points,
+        )
+
+        # Set Gaussian Process parameters
+        optimizer.set_gp_params(alpha=self.alpha, n_restarts_optimizer=self.n_restarts)
+
+        # Use initial point as first probe
+        probe_params = {f"param{i}": float(val) for i, val in enumerate(x0)}
+        optimizer.probe(
+            params=probe_params,
+            lazy=True,
+        )
+        optimizer.maximize(
+            init_points=self.init_points,
+            n_iter=self.n_iter,
+        )
+
+        res = _process_bayes_opt_result(optimizer=optimizer, x0=x0, problem=problem)
+        return res
+
+    def _process_constraints(
+        self, constraints: list[dict[str, Any]] | None
+    ) -> NonlinearConstraint | None:
+        """Temporarily skip processing of nonlinear constraints.
+
+        Args:
+            constraints: List of constraint dictionaries from the problem
+
+        Returns:
+            None. Nonlinear constraint processing is deferred.
+
+        """
+        # TODO: Implement proper handling of nonlinear constraints in future.
+        return None
+
+
+def _process_bounds(bounds: InternalBounds) -> dict[str, tuple[float, float]]:
+    """Process bounds for bayesian optimization.
+
+    Args:
+        bounds: Internal bounds object.
+
+    Returns:
+        Dictionary mapping parameter names to (lower, upper) bound tuples.
+
+    Raises:
+        ValueError: If bounds are None or infinite.
+
+    """
+    if not (
+        bounds.lower is not None
+        and bounds.upper is not None
+        and np.all(np.isfinite(bounds.lower))
+        and np.all(np.isfinite(bounds.upper))
+    ):
+        raise ValueError(
+            "Bayesian optimization requires finite bounds for all parameters. "
+            "Bounds cannot be None or infinite."
+        )
+
+    return {
+        f"param{i}": (lower, upper)
+        for i, (lower, upper) in enumerate(zip(bounds.lower, bounds.upper, strict=True))
+    }
+
+
+def _extract_params_from_kwargs(
+    kwargs: dict[str, Any], n_params: int
+) -> NDArray[np.float64]:
+    """Extract parameters from kwargs dictionary.
+
+    Args:
+        kwargs: Dictionary with parameter values.
+        n_params: Number of parameters to extract.
+
+    Returns:
+        Array of parameter values.
+
+    """
+    return np.array([kwargs[f"param{i}"] for i in range(n_params)])
```

Comment:
Wouldn't `np.array(list(kwargs.values))` do the same without requiring a user to pass `n_params`?

Also: `kwargs` is not a good name here. Maybe use `params_dict` instead?